### PR TITLE
Fix wlog.Watcher increased replay time on startup

### PR DESCRIPTION
**Pull request by jlevesy**
### What does this PR do?

- ebce11de449c8d79a8b2e4e361c762f887f54e77 Makes the `agent.DB` notify the remote storage (via a `wlog.Notifier`) when a commit happens.
- eb73a8e19e7aa83b9a4e044ac5aa14979335bca5 Makes the `wlog.Watcher.watch` method read a segment synchronously when it is not tailing the last segment instead of waiting. It includes a test that makes sure we're able to replay within a single timeout interval a generated set of segments. I'm not exactly sure that's what you expected, but I can't see how a real benchmark would help here.

Those two fixes commit are the outcome of the discussion happening in https://github.com/prometheus/prometheus/issues/13111

Fixes https://github.com/prometheus/prometheus/issues/13111

### CODE REVIEW

**Comment by jlevesy** [comment]
I'll roll out that branch on my dev environment on Monday, see if it makes the expected changes :)

**Comment by bboreham** [comment]
This feels like two PRs to me: one allowing the agent to participate in Notify calls, and one changing the non-tail behaviour to not need a Notify.

If I'm mistaken, please modify the description to highlight how they are related.

**Comment by jlevesy** [comment]
> This feels like two PRs to me: one allowing the agent to participate in Notify calls, and one changing the non-tail behaviour to not need a Notify.

It is indeed. It was the outcome of the discussion happening in #13111 so I packed them into a single PR. But I can open two separate PRs if you want!

File: tsdb/wlog/watcher.go [hunk 1]
```diff
@@ -487,6 +485,8 @@ func (w *Watcher) watch(segmentNum int, tail bool) error {
 			readTicker.Reset(readTimeout)
 
 		case <-w.readNotify:
+			level.Debug(w.logger).Log("msg", "Watcher is reading the WAL due to notify")
```
[the remark this answers is not in the document]
Comment (jlevesy) [reply tsdb/wlog/watcher.go:488]: Not necessarily, that's a debugging leftover sorry. Will clean up!

**Comment by bboreham** [comment]
I find it easier to think about self-contained changes, plus we get the benefit of seeing that CI passes on each independent change.

**Comment by cstyan** [comment]
I don't have a preference either way; these are relatively small changes and you've already isolated them nicely in separate commits.

**Comment by mattdurham** [comment]
Looks reasonable from my standpoint. The only edge case I see is abusing to SetWriteNotified could lead to a nil panic but that seems like an abuse and likely not worth the mutex.

**Comment by cstyan** [comment]
@mattdurham fair point, we should probably at least have a comment on the interface that `SetWriteNotified` is only meant to be called once/during initialization. We can set up more safety around it's usage if we find a need to, a lock has some cost. Thanks for taking a look :+1:

**Comment by jlevesy** [comment]
Opened #13223 and #13224. Closing this PR.

**Pull request closed without merging**